Cleaning system

TasnimAwashreh opened the pull request.

TasnimAwashreh committed:
- Added database methods
- Added repositories with their interfaces
- Added services with their interfaces
- Users can be two types: passanger and manager
- Added passenger commands as enums
- Added manager commands as enums
- Added handlers for the commands
- Flights can be validated and uploaded by the manager

TasnimAwashreh committed:
- Added PresentAndFutureOnly attribute for DepartureDate
- Forgot to add conn.Close() to functions that use connections
- Corrected some names

TasnimAwashreh committed:
- renamed namespaces
- removed whitespace
- updated some strings

RandaZraik commented on AirportTicketBookingExercise/Data/Attributes/PresentAndFutureOnlyAttribute.cs:
```diff
@@ -0,0 +1,33 @@
+namespace ATB.Data.Attributes
+{
+    [System.AttributeUsage(System.AttributeTargets.Field | System.AttributeTargets.Parameter | System.AttributeTargets.Property, AllowMultiple = false)]
+    public class PresentAndFutureOnlyAttribute : ValidationAttribute
```
where are you using this attribute?

TasnimAwashreh replied:
This is used for the field DepartureDate in the model Flight. It is used to check during the flights csv upload since the DepartureDate should not be older than today.

RandaZraik commented on AirportTicketBookingExercise/Data/BookingFilter.cs:
```diff
@@ -0,0 +1,17 @@
+
+namespace ATB.Data
+{
+    public class BookingFilter
```
This shouldn't be in the Data layer, logic?

RandaZraik commented on AirportTicketBookingExercise/Logic/Enums/BookingClass.cs:
```diff
@@ -0,0 +1,25 @@
+{
+    public enum BookingClass
+    {
+        first,
```
Enum values should be in PasalCase. Also add explicit integers (None = 0, First = 1, ...) to prevent breaking changes when enum members are reordered, ensure consistent database storage and API serialization.

RandaZraik commented on AirportTicketBookingExercise/Logic/Enums/BookingClass.cs:
```diff
@@ -0,0 +1,25 @@
+    }
+    public class BookingClasses
+    {
+        public static BookingClass strToBookingClass(string bookingClass)
```
PascalCase. `ToBookingClass`

RandaZraik replied:
you make this an extesnsion method

RandaZraik commented on AirportTicketBookingExercise/Logic/Enums/BookingFilters.cs:
```diff
@@ -0,0 +1,103 @@
+
+    public enum FilterParam
+    {
+        flight,
```
same comment about assigning a integer to each value.

RandaZraik commented on AirportTicketBookingExercise/App/Program.cs:
```diff
@@ -0,0 +1,56 @@
+class Program
+{
+
+    public static void startLoop(IServiceProvider serviceProvider)
```
Lets update all public methods in this PR to PascalCase i.e. `StartLoop`.

RandaZraik commented on AirportTicketBookingExercise/Data/Db/DAOs/BookingDao.cs:
```diff
@@ -0,0 +1,111 @@
+
+        public void CreateBooking(Booking booking)
+        {
+            using var conn = new SqliteConnection($"Data Source={_connectionString}");
```
The task is about using a file system not a database. Lets please re-work the PR and use file systems instead of the databases :D

RandaZraik replied:
Once this change done, I will review again.

RandaZraik commented on AirportTicketBookingExercise/Data/Db/DAOs/BookingDao.cs:
```diff
@@ -0,0 +1,111 @@
+
+namespace ATB.Data.Db.DAOs
+{
+    public class BookingDAO
```
I don't see any difference between DAOs and Repositories, both handle db/data operations. Lets please get rid off DAO's and just use instead repositories. Repositories are meant to hold the CRUD operations against DB/Data.

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/ManagerCommandHandler.cs:
```diff
@@ -0,0 +1,166 @@
+
+namespace ATB.Logic.Handlers.Command
+{
+    public class ManagerCommandHandler
```
Can you please move all classes and logic related to UI folder (get input from user and output messages).

RandaZraik replied:
So we have clear separations:
- Data Layer
- Logic Layer
- UI Layer

Make sure please each file belongs to the right layer and folder :)

TasnimAwashreh committed:
- Enums are now in Pascal Case

TasnimAwashreh committed:
- Public method and field names are now in pascal case
- Made sure that private variables are in camel case
- Clarified method names

TasnimAwashreh committed:
- BookingClass, BookingFilter, Manager + Passenger commands have been changed into extension methods

TasnimAwashreh committed:
- Removed DAOs

TasnimAwashreh committed:
- Handlers no longer prints out to console
- BookingManager has been moved to App section to keep inputs+outputs seperate from Logic layer

TasnimAwashreh committed:
- Moved BookingFilter to Logic

TasnimAwashreh committed:
- Application now uses a file system

TasnimAwashreh committed:
- Application now uses a file system

RandaZraik commented on AirportTicketBookingExercise/App/BookingManager.cs:
```diff
@@ -0,0 +1,194 @@
+{
+    public class BookingManager
+    {
+
```
please remove extra newline

RandaZraik replied:
please apply to all files

RandaZraik commented on AirportTicketBookingExercise/Logic/Enums/BookingClass.cs:
```diff
@@ -0,0 +1,25 @@
+        Business = 2,
+        Economy = 3
+    }
+    public static class BookingClasses
```
add a newline before this please

RandaZraik replied:
please apply to all files

RandaZraik commented on AirportTicketBookingExercise/Logic/Enums/BookingClass.cs:
```diff
@@ -0,0 +1,25 @@
+﻿
+
```
remove extra two lines

RandaZraik commented on AirportTicketBookingExercise/Logic/Enums/BookingFilters.cs:
```diff
@@ -0,0 +1,103 @@
+
+namespace ATB.Logic.Enums
+{
+
```
extra newline, please review :D lets do this for all files in the PR, try to review files yourself one by one :)

RandaZraik commented on AirportTicketBookingExercise/AirportTicketBookingExercise.csproj:
```diff
@@ -7,4 +7,9 @@
 
+  <ItemGroup>
+    <PackageReference Include="CsvHelper" Version="33.1.0" />
+    <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="9.0.8" />
```
please remove as it's no longer used

RandaZraik commented on AirportTicketBookingExercise/Configuration/ServiceCollectionExtensions.cs:
```diff
@@ -0,0 +1,59 @@
+        public static IServiceCollection AddRepositories(this IServiceCollection services)
+        {
+            const string bookingFile = "BookingDb.db";
+            const string userFile = "BookingDb.db";
```
why the user file is the same as the booking file?

RandaZraik commented on AirportTicketBookingExercise/Data/Db/DatabaseManager.cs:
```diff
@@ -0,0 +1,45 @@
+            _usersPath = usersPath;
+            _bookingsPath = bookingsPath;
+        }
+        public void CreateDatabase()
```
this should talk list of paths, and the following is duplicate , it can be abstracted and made generic where it takes any path and create the db for it, use generics `<TModelMap>`

RandaZraik replied:
shouldn't we also create a db for flights?

RandaZraik commented on AirportTicketBookingExercise/Data/Models/User.cs:
```diff
@@ -0,0 +1,32 @@
+{
+    public enum UserType
+    {
+        Manager,
```
please assign an integer to each value

RandaZraik commented on AirportTicketBookingExercise/Data/Repository/UserRepository.cs:
```diff
@@ -0,0 +1,73 @@
+    {
+        private readonly string _usersPath;
+
+        public UserRepository(string usersPath) {
```
please put { on a newline

RandaZraik commented on AirportTicketBookingExercise/Data/Repository/BookingRepository.cs:
```diff
@@ -0,0 +1,125 @@
+            _bookingPath = bookingPath;
+        }
+
+        private Booking MapBooking(SqliteDataReader reader)
```
do we still need this method?

RandaZraik commented on AirportTicketBookingExercise/Data/Repository/FlightRepository.cs:
```diff
@@ -0,0 +1,143 @@
+    public class FlightRepository: IFlightRepository
+    {
+        private readonly string _flightsCSVPath;
+        private List<Flight> _flights;
```
we don't need this. we are working directly with the csv file

RandaZraik commented on AirportTicketBookingExercise/Data/Repository/FlightRepository.cs:
```diff
@@ -0,0 +1,143 @@
+                        strBuilder.Append($"Row {rowCount}: {attr.ErrorMessage} \n");
+                }
+            }
+            catch (Exception e) {
```
newline before `{`

RandaZraik commented on AirportTicketBookingExercise/Data/Repository/FlightRepository.cs:
```diff
@@ -0,0 +1,143 @@
+            flight.SeatsAvailable--;
+        }
+
+        private void ValidateField<T>(string fieldName, string fieldInput, 
```
these validation methods are business logic that shouldn't be in the repo, should be moved to the service level. Repos are only for direct CRUD operations against DB.

RandaZraik commented on AirportTicketBookingExercise/Data/Repository/FlightRepository.cs:
```diff
@@ -0,0 +1,143 @@
+            }
+        }
+
+        public bool ImportCSVData()
```
the requirement is that we need to import flights from an external csv file (which is not a db really), this method should expect a file path, and this is business logic should be moved to the service layer. 

This of this as a service method that we can use to import data from a third party storage to our system. Regardless if both are using csv storage

RandaZraik commented on AirportTicketBookingExercise/Data/Repository/FlightRepository.cs:
```diff
@@ -0,0 +1,143 @@
+                csvParser.ReadLine();
+                StringBuilder strBuilder = new StringBuilder("");
+                int rowCount = 1;
+                while (!csvParser.EndOfData)
```
instead of validating members this way, you can just load all flights into list of models and you need to use the data annotations you assigned to each property, you can validate the model using `Validator` provided by `DataAnnotations` library.

RandaZraik commented on AirportTicketBookingExercise/Data/Repository/FlightRepository.cs:
```diff
@@ -0,0 +1,143 @@
+                csvParser.ReadLine();
+                try
+                {
+                    while (!csvParser.EndOfData)
```
why we are not using `csv.GetRecords`

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/BookingService.cs:
```diff
@@ -0,0 +1,117 @@
+        {
+            var Bookings = _bookingRepository.GetAllBookings();
+            var Flights = _flightRepository.GetFlights();
+            var passengers = _userRepository.GetAllUsers().Where(u => u.UserType == UserType.Passenger).ToList();
```
the filtration should happen on the db side, we shouldn't return all records and then do the filtration here in general, lets add a new repo method to get users by type

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/BookingService.cs:
```diff
@@ -0,0 +1,117 @@
+            var Flights = _flightRepository.GetFlights();
+            var passengers = _userRepository.GetAllUsers().Where(u => u.UserType == UserType.Passenger).ToList();
+
+            var Filtered = Bookings.Where(b =>
```
all filtration logic should happen in the repo method, introduce a new method for filtration there

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/BookingService.cs:
```diff
@@ -0,0 +1,117 @@
+
+            var Filtered = Bookings.Where(b =>
+            {
+                var flight = Flights.FirstOrDefault(f => f.FlightId == b.FlightId);
```
`where` is a loop and here you are doing another looping `FirstOrDefault`. think about using dictionary.

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/BookingService.cs:
```diff
@@ -0,0 +1,117 @@
+            var Filtered = Bookings.Where(b =>
+            {
+                var flight = Flights.FirstOrDefault(f => f.FlightId == b.FlightId);
+                var passenger = passengers.FirstOrDefault(p => p.UserId == b.PassengerId);
```
same comment as above

RandaZraik commented on AirportTicketBookingExercise/App/BookingManager.cs:
```diff
@@ -0,0 +1,194 @@
+            else Console.WriteLine("\n Please enter an appropriate action");
+        }
+
+        public void executeManagerCommand(string[] productInfo, ManagerCommand command)
```
PascalCase `ExecuteManagerCommand`

RandaZraik commented on AirportTicketBookingExercise/App/Program.cs:
```diff
@@ -0,0 +1,54 @@
+        }
+    }
+
+    public static string introduction()
```
PascalCase: `Introduction`

RandaZraik commented on AirportTicketBookingExercise/App/BookingManager.cs:
```diff
@@ -0,0 +1,194 @@
+        public void ProcessInput(string input)
+        {
+            string[] line = input.Split(' ');
+            ManagerCommand managerCommand = line[0].ParseManagerCommand();
```
if input is empty, `line[0]` will throw an exception

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/ManagerCommandHandler.cs:
```diff
@@ -0,0 +1,84 @@
+
+namespace ATB.Logic.Handlers.Command
+{
+    public class ManagerCommandHandler
```
isn't this just a `ManagerService`

RandaZraik replied:
@TasnimAwashreh i see you have moved handlers to `App` layer, but these are more like services with business logic, why you think they should be in `App`? still need to be renamed and moved to the `Logic` layer. We should never have business logic in the `App` layer.

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/ManagerCommandHandler.cs:
```diff
@@ -0,0 +1,84 @@
+        }
+        public string Validate()
+        {
+            Console.WriteLine($"\n Validating... \n");
```
this should be in App layer

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/ManagerCommandHandler.cs:
```diff
@@ -0,0 +1,84 @@
+            return bookingResults;  
+        }
+
+        public User? ManagerLogIn(string[] productInfo, User? loggedInUser)
```
this service shouldn't understand `user inputs`, should just pass username and password here

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/ManagerCommandHandler.cs:
```diff
@@ -0,0 +1,84 @@
+            return null;
+        }
+
+        public bool ManagerSignUp(string[] productInfo)
```
same comment as above

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/ManagerCommandHandler.cs:
```diff
@@ -0,0 +1,84 @@
+        {
+            if (productInfo.Length < 3)
+                return false;
+            else
```
no need for the else, `return false` will do the job and not continue in the method

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+
+namespace ATB.Logic.Handlers.Command
+{
+    public class PassengerCommandHandler
```
isn't this just a `PassengerService`

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+            _flightService = flightService;
+        }
+
+        public bool PassengerBookFlight(string[] productInfo, User loggedInUser)
```
same comment about services shouldn't know about UI and expect to have actual info passed to it, as this can be called from API layer, CMD wrapper, ... different inputs but one service

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+                Flight? flight = _flightService.GetFlight(flightId);
+                if (flight == null || flight.SeatsAvailable >= flight.SeatCapacity)
+                    return false;
+                else
```
no need for else, return on line 31 is enough

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+                    {
+                        var Booking = new Booking
+                        {
+                            BookingId = new Random().Next(100000, 999999),
```
generating ids should be on the db level

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/ManagerCommandHandler.cs:
```diff
@@ -0,0 +1,84 @@
+                {
+                    var user = new User
+                    {
+                        UserId = new Random().Next(100000, 999999),
```
generating ids should be on the db level. Also move this to an extension method to be used for other entities

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+            }
+        }
+
+        public List<Flight> Search(string[] productInfo)
```
same comment, no pass for generic user inputs

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+            return filteredFlights;
+        }
+
+        public bool Cancel(string[] productInfo, User loggedInUser)
```
same comment, no pass for generic user inputs

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+                int bookingId = int.Parse(productInfo[1]);
+                if (!_bookingService.IsBookingValidById(bookingId, loggedInUser.UserId))
+                    return false;
+                else
```
no  need for else, return on line 101 is enough

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+            }
+            catch
+            {
+                return false;
```
we return false in multiple paths here, each one for a different reason, we shouldn't swallow exception and just return false, better to throw an  excpetion with the proper message for each path that we think is failing.

RandaZraik replied:
re-review the PR and see where we return false and if we should throw exceptions instead

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+        {
+            if (productInfo.Length < 3)
+                return false;
+            else
```
no need for else

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+            }
+        }
+
+        public bool Modify(string[] productInfo, User loggedInUser)
```
same comment, no pass for generic user inputs

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+            
+        }
+
+        public bool PassengerSignUp(string[] productInfo)
```
same comment, no pass for generic user inputs

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+
+        }
+
+        public User? PassengerLogIn(string[] productInfo, User? loggedInUser)
```
same comment, no pass for generic user inputs

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+        {
+            if (productInfo.Length < 3)
+                return false;
+            else
```
no need for else

RandaZraik commented on AirportTicketBookingExercise/Logic/Handlers/Command/PassengerCommandHandler.cs:
```diff
@@ -0,0 +1,237 @@
+        {
+            if (loggedInUser == null)
+                return false;
+            else
```
no need for else

TasnimAwashreh committed:
- Removed extra newlines
- Added newlines where needed
- Fixed CSV paths

TasnimAwashreh committed:
- Removed/Added extra new lines pt.2
- Added integers to UserType enum

TasnimAwashreh committed:
- Moved some methods to the Service area to keep the repository area CRUD related only

TasnimAwashreh committed:
- Flights now has its own file
- Upload saves data into Flights file if data is valid
- Upload and Validate now requires a path to the csv file of your choice

TasnimAwashreh committed:
.

TasnimAwashreh committed:
- Removed extra 'elses'

TasnimAwashreh committed:
- Moved handlers to App
- Removed generic input in the handlers

TasnimAwashreh committed:
- Added messages to exceptions

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/BookingService.cs:
```diff
@@ -0,0 +1,120 @@
+
+        public List<Booking> FilterBookings(BookingFilter query)
+        {
+            var Bookings = _bookingRepository.GetAllBookings();
```
Local variables should be camalCase: `bookings`

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/BookingService.cs:
```diff
@@ -0,0 +1,120 @@
+        public List<Booking> FilterBookings(BookingFilter query)
+        {
+            var Bookings = _bookingRepository.GetAllBookings();
+            var Flights = _flightRepository.GetFlights();
```
same comment as above

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/UserService.cs:
```diff
@@ -0,0 +1,52 @@
+    public class UserService : IUserService
+    {
+        private readonly IUserRepository _userRepository;
+        public UserService(IUserRepository userRepository) { 
```
review all files where is `{` misplaced and fix please :D

TasnimAwashreh committed:
- creating Id for users and booking has moved into the Data area

TasnimAwashreh committed:
- Formatted code
- Fixed variables that had pascal case to camel case

TasnimAwashreh committed:
- Updated method names

TasnimAwashreh committed:
Removed unused space

TasnimAwashreh committed:
- DatabaseManager uses a more generic method

TasnimAwashreh committed:
- Changed the flight validation from checking one field at a time to validating a list

TasnimAwashreh committed:
- Updated filter to use dictionary to increase performance

TasnimAwashreh committed:
- Removed Handlers
- Methods that were in handler are now in Services

RandaZraik commented on AirportTicketBookingExercise/Data/Db/DatabaseManager.cs:
```diff
@@ -0,0 +1,51 @@
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine($"Please fix Create Database error (below) and restart the system. \nError: {ex}");
```
we shouldn't print things here, lets the exception bubble up and called to handle the exception and print messages to the users

RandaZraik commented on AirportTicketBookingExercise/Data/Repository/BookingRepository.cs:
```diff
@@ -0,0 +1,124 @@
+            List<Booking> records = new List<Booking>();
+            try
+            {
+                using (var reader = new StreamReader(_bookingPath))
```
we have this logic that reads from csv files in each repo, too much duplicates, lets apply DRY principle and have a csv helper with the main operations you would do with a csv file: read, write , ... and this update all repos to use this helper instead.

RandaZraik commented on AirportTicketBookingExercise/Data/Repository/BookingRepository.cs:
```diff
@@ -0,0 +1,124 @@
+            booking.GenerateBookingId();
+            try
+            {
+                using (var writer = new StreamWriter(_bookingPath, append: true))
```
same comment as above

RandaZraik commented on AirportTicketBookingExercise/Data/Repository/BookingRepository.cs:
```diff
@@ -0,0 +1,124 @@
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine(ex.ToString());
```
we shouldn't be printing anything from repos and services (data and logic)  layers. As mentioned before, App/UI layer is the one calling the data/logic and prints to the screen, We should have data and logic layers abstracted and just doing what it should do, so we can use this repo from other areas where we might not have CMD to print things to screen.

when you work later with APIs, we don't have CMD, we return data with APIs. so we shouldn't be expecting console.writeline here.

also here we are swallowing the exception, the exception should bubble up and lets the caller to handle the exception , as it's a generic exception.

RandaZraik replied:
please, review all files here and apply this comment

RandaZraik commented on AirportTicketBookingExercise/Logic/Extensions/IdGenerator.cs:
```diff
@@ -0,0 +1,18 @@
+
+namespace ATB.Logic.Extensions
+{
+    public static class IdGenerator
```
this should be in the data layer, why logic?

RandaZraik commented on AirportTicketBookingExercise/Logic/Enums/PassengerCommands.cs:
```diff
@@ -0,0 +1,47 @@
+﻿namespace ATB.Logic.Enums
+{
+    public enum PassengerCommand
```
isn't this should used in App layer? if so, it should be moved to that layer

RandaZraik commented on AirportTicketBookingExercise/Logic/Enums/ManagerCommands.cs:
```diff
@@ -0,0 +1,38 @@
+﻿namespace ATB.Logic.Enums
+{
+    public enum ManagerCommand
```
isn't this should used in App layer? if so, it should be moved to that layer

TasnimAwashreh committed:
- Commands have been moved to the App layer
- Moved Id Generator to Logic
- Repos and Services no longer print out anything so that the exception bubbles up to the App layer

TasnimAwashreh committed:
- Added a csv helper class containing generic methods so that any repository can use the code

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/BookingService.cs:
```diff
@@ -0,0 +1,174 @@
+            var flights = _flightRepository.GetFlights().ToDictionary(f => f.FlightId);
+            var passengers = _userRepository.GetUsersByType(UserType.Passenger).ToDictionary(u => u.UserId);
+
+            bookingResults = bookings.Where(b =>
```
think about rewriting this using joins and just one straightforward query, is a bit odd to in the where statement all this logic, can be done with multiple where statements

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/BookingService.cs:
```diff
@@ -0,0 +1,174 @@
+            }
+            catch (Exception ex)
+            {
+                Console.WriteLine($"Error while booking: {ex.ToString()}");
```
as mentioned before, we shouldn't have print statements in logic layer please.

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/FlightService.cs:
```diff
@@ -0,0 +1,164 @@
+                }
+                catch (Exception ex)
+                {
+                    Console.WriteLine($"Error while importing flight data: {ex.ToString()}");
```
same comment as above

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/UserService.cs:
```diff
@@ -0,0 +1,69 @@
+            }
+            catch (Exception ex) 
+            {
+                Console.WriteLine($"Error creating a user: {ex}");
```
same comment as above, please apply comments you receive in one place across the whole PR ;)

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/BookingService.cs:
```diff
@@ -0,0 +1,174 @@
+                bool isSuccess = _bookingRepository.DeleteBooking(bookingId);
+                if (!isSuccess)
+                    return false;
+                else return true;
```
no need for else

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/UserService.cs:
```diff
@@ -0,0 +1,69 @@
+                    UserType = usertype
+                };
+                if (_userRepository.GetUser(name) != null)
+                    return false;
```
we should throw a validation exception

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/UserService.cs:
```diff
@@ -0,0 +1,69 @@
+                if (_userRepository.GetUser(name) != null)
+                    return false;
+                _userRepository.CreateUser(user);
+                return true;
```
no need, exceptions better be thrown if validations are failing. true/false doesn't indicate much what went wrong

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/UserService.cs:
```diff
@@ -0,0 +1,69 @@
+            {
+                return user;
+            }
+            return null;
```
same comment about throwing exception. We should throw an unauthenticated exception

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/FlightService.cs:
```diff
@@ -0,0 +1,164 @@
+
+        }
+
+        public List<Flight> Search(string[] filterInfo)
```
filterinfo is pretty generic, we should be strongly types and be specific of this method should expect, with names params

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/FlightService.cs:
```diff
@@ -0,0 +1,164 @@
+            string input = filterInfo[2];
+
+            var flights = _flightRepo.GetFlights();
+            filteredFlights = SearchParam switch
```
filtration should happen in the repo level,

RandaZraik replied:
we can filter by multiple fields, so can have nested where statements, so switch here won't be enough

RandaZraik commented on AirportTicketBookingExercise/App/Utils/CsvActionsHelper.cs:
```diff
@@ -0,0 +1,67 @@
+
+namespace AirportTicketBookingExercise.App.Utils
+{
+    public class CsvActionsHelper
```
this is not an App related helper, should be in logic. I don't UI cares about this

RandaZraik commented on AirportTicketBookingExercise/Data/Db/DatabaseManager.cs:
```diff
@@ -0,0 +1,51 @@
+            }
+        }
+
+        private void CreateCSVFile<TModel, TMap>(string path)
```
can't we use CsvActionsHelper?

TasnimAwashreh committed:
- Moved Helper to Logic
- Moved CreateDatabase to Helper

TasnimAwashreh committed:
- All filtering now happens in the repository

TasnimAwashreh committed:
- Fixed search parameters to less generic ones
- Moved search filter to repository
- Removed unnecessay returns, print outs

TasnimAwashreh committed:
- Added attributes to User

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/UserService.cs:
```diff
@@ -0,0 +1,69 @@
+            }
+
+            if (_userRepository.GetUser(name) != null)
+                return false;
```
this is a validation issue, we should fail and throw an exception so the user knows that the name already used

RandaZraik commented on AirportTicketBookingExercise/Logic/Service/BookingService.cs:
```diff
@@ -0,0 +1,122 @@
+
+            bookingResults =
+                (
+                 from b in bookings
```
this is more of a db query that should be moved to a repo, along with other retrieval statements above, a repo method to get and join bookings and flights and receives the passenger userid as a parameter

TasnimAwashreh committed:
- Removed !csvParser.EndOfData

TasnimAwashreh committed:
- added an if statement to check if the input is null
- moved filter to its own repo
- updated search to search to use linq rather than a switch statement
- added throw exceptions if the user did not enter the correct password or if username already exists during signup

TasnimAwashreh committed:
- Refactored App layer
- Moved executing commands methods to CommandExecuter, where it will contain the switch statements, and Helper files to prevent clutter in the switch statements

TasnimAwashreh committed:
pt 2

RandaZraik commented on AirportTicketBookingExercise/Domain/BookingFilter.cs:
```diff
@@ -0,0 +1,15 @@
+﻿namespace ATB.Logic
+{
+    public class BookingFilter
```
lets have it in models  in domain

RandaZraik commented on AirportTicketBookingExercise/Domain/Service/FlightService.cs:
```diff
@@ -0,0 +1,111 @@
+
+        public List<Flight> GetFlights()
+        {
+            List<Flight> flights = new List<Flight>();
```
no need to initialize a new list, just return GetFlights directly

RandaZraik commented on AirportTicketBookingExercise/Domain/Service/FlightService.cs:
```diff
@@ -0,0 +1,111 @@
+        {
+            var strBuilder = new StringBuilder("");
+            int rowCount = 1;
+            var flights = new List<Flight>();
```
this initializes new list, no need. you can just do: ```var flights = CsvActionsHelper.GetAllRecords<Flight, FlightMap>(importPath);```

RandaZraik commented on AirportTicketBookingExercise/Domain/Service/FlightService.cs:
```diff
@@ -0,0 +1,111 @@
+            return true;
+        }
+
+        private Flight? AddFlight(string[] fields)
```
not used please remove

RandaZraik commented on AirportTicketBookingExercise/Domain/Service/BookingService.cs:
```diff
@@ -0,0 +1,110 @@
+        {
+            Flight? flight = _flightRepository.GetFlight(flightId);
+            if (flight == null || flight.SeatsAvailable >= flight.SeatCapacity)
+                return false;
```
return false in each path here should be an exception. These are more of validation error

RandaZraik commented on AirportTicketBookingExercise/Domain/Service/BookingService.cs:
```diff
@@ -0,0 +1,110 @@
+        public bool Modify(int bookingId, BookingClass bookingClass, User loggedInUser)
+        {
+            if (!_bookingRepository.IsBookingValidById(bookingId, loggedInUser.UserId))
+                return false;
```
same thing, should be a validation exception

RandaZraik replied:
lets check other methods and update please

TasnimAwashreh committed:
- removed boolean returns and replaced them with throwing exceptions
- Moved bookingfilter to its own Models folder
-Removed unnecessary initializations

TasnimAwashreh committed:
- Fixed service name
- Fixed some methods to take in the model rather than indiviual parameters
- Updated Exceptions
